### DISCUSSION TITLE
Windows 11 Style Icon

### DISCUSSION THREAD

**Discussion by D-Arora** [File-New-Project/EarTrumpet#1077] [Ideas]
Could you please add the option for a Windows 11 style icon. Thanks! Love your work!

**Comment by riverar** [comment]
Can you elaborate on what you're referring to as "Windows 11 style icon"? Thanks!

**Reply by D-Arora** [reply]
Something more in line with the audio icon in the Windows 11 quick settings menu. The icon is a little bit more compact and has thicker line weights. Also when reducing volume the sound lines disappear instead of fading to grey. 

![image](https://user-images.githubusercontent.com/67812826/155416893-47fa5dd2-1a33-463c-a9b8-734ca707a102.png)

**Reply by D-Arora** [reply]
Hi @riverar,

I have extracted the Windows 11 audio icons as icos and svgs (Attached [here](https://github.com/File-New-Project/EarTrumpet/files/8161325/W11_Audio.zip)).

I was trying to work on a fork of EarTrumpet to make the icons mod but I keep getting build errors in Visual Studio. Have you encountered this error before? :

`EarTrumpet\EarTrumpet.ColorTool\MainWindow.xaml(18,18): error MC3074: The tag 'Binding' does not exist in XML namespace 'clr-namespace:EarTrumpet.Extensions.EventBinding;assembly=EarTrumpet'. Line 18 Position 18.`

Thanks

**Reply by riverar** [reply]
@D-Arora Haven't seen that -- are you using Visual Studio 2019? .NET Framework / WPF support installed? See also https://github.com/File-New-Project/EarTrumpet/blob/master/COMPILING.md

Appreciate your research, one thing to note: We cannot embed any assets from Windows into the app for copyright reasons. We can, however, load them at runtime from platform libraries/resources. Thanks!

**Reply by D-Arora** [reply]
I'm using Visual Studio 2022 and have .NET and the Windows SDK installed. I'll try downgrading to 2019 maybe that might help.

In regards to the assets, what if they were licensed from somewhere like icons8 instead of Windows? I checked and they do have the icons available for use for free if under 100px x 100px - https://icons8.com/icon/set/media-controls/fluency-systems-regular. 

Their licensing agreement is also here - https://icons8.com/vue-static/landings/pricing/icons8-license.pdf

**Reply by D-Arora** [reply]
I've downgraded to Visual Studio 2019 and have a debug build working with the changes - https://github.com/D-Arora/EarTrumpet
Still can't get a release build working though :(
I'm getting a similar issue:

```Build started...
1>------ Build started: Project: EarTrumpet, Configuration: Release x86 ------
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2203,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Windows". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors.
1>C:\Users\iamdh\Documents\Work\FOSS\D-Arora\EarTrumpet\packages\GitVersionTask.5.5.1\build\GitVersionTask.targets(19,9): warning : File contains assembly version attributes which conflict with the attributes generated by GitVersion obj\x86\Release\AssemblyInfo.g.cs
1>C:\Users\iamdh\Documents\Work\FOSS\D-Arora\EarTrumpet\EarTrumpet\Interop\Helpers\PackageHelper.cs(2,15,2,31): error CS0234: The type or namespace name 'ApplicationModel' does not exist in the namespace 'Windows' (are you missing an assembly reference?)
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
========== Deploy: 0 succeeded, 0 failed, 0 skipped ==========
```

**Reply by D-Arora** [reply]
Nevermind the release build works now. I had to remove the references to System.Runtime.WindowsRuntime and Windows and add them back again manually.

**Reply by D-Arora** [reply]
This is it working:


https://user-images.githubusercontent.com/67812826/156463762-55e59387-7fe6-4117-8568-77e1ffa153a6.mp4


I'm going to continue working on this fork. A few changes I'd like to implement are:

- Change the sliders to match the new [WinUI slider style](https://docs.microsoft.com/en-us/windows/apps/design/controls/slider)
- Changing the flyout margins to be closer to the Win11 quick settings menu
- Match the acrylic styling to Win11, removing the accent colour subsections

Reach goals:
- Adding sound mix presets that move specific apps to certain outputs
- Changing the system sounds icon to be something more modern
- Adding the option to hide the system sounds icon and subsequently allowing the flyout to autohide devices with no 'active' apps
- Changing icons for different outputs as per #1011 

If anyone has any other ideas I would be happy to give them a try.

**Reply by riverar** [reply]
> Change the sliders to match the new [WinUI slider style](https://docs.microsoft.com/en-us/windows/apps/design/controls/slider)
> Changing the flyout margins to be closer to the Win11 quick settings menu
> Match the acrylic styling to Win11, removing the accent colour subsections

These sound great, will be very happy to accept/review those. (Perhaps it's time we plumb in modernwpf or another similar library? Will have to work with our existing theming logic...)

> Adding sound mix presets that move specific apps to certain outputs

Believe we already support moving apps on start to devices via EarTrumpet Actions.

> Changing the system sounds icon to be something more modern

We want to blend in with Windows and not stand out, so preferably we'd continue to use platform icons. We are not particularly interested in custom icons but I understand Windows 11 is applying pressure on us to do this... will think some more about this.

> Adding the option to hide the system sounds icon and subsequently allowing the flyout to autohide devices with no 'active' apps

These sound easier/do-able but be aware this incurs a localization hit/tax.

> Changing icons for different outputs as per https://github.com/File-New-Project/EarTrumpet/discussions/1011

Hm, will discuss this with the team but I don't think we want to stray too far from what the OS shows. EarTrumpet is currently wired up into OS infrastructure and this is the icon they're telling us to use. Straying from that could result in unexpected behavior. (That said, I suppose this could ship off-by-default. Just be aware this introduces a bunch of settings complexity.)

**Reply by riverar** [reply]
Also make sure you're on dev for those changes, your video indicates to me you're not on dev (missing padding from the edge of the display).

**Reply by D-Arora** [reply]
Thanks for the feedback @riverar, I definitely agree with the blending into windows rule. I think limiting the customisable icons to windows dll/system icons would be the way to go. That said some of the legacy windows icons (like system sounds) just haven't been restyled yet so I might make some alternatives that have updated styles.

P.S. By dev do you mean the dev branch? Or a debug build?

**Reply by riverar** [reply]
`dev` branch, sorry for the confusion.

**Reply by D-Arora** [reply]
I thought of some additional functionality + questions and wanted your feedback.

1. [Feature] As a new feature, has the dev team considered the ability to drag and drop apps between audio devices
2. [Feature] Renaming audio devices. Use case is when connected to docks/displays which have long model numbers as output names.
3. How do you access Eartrumpet Actions? Does it just move certain apps to specific outputs on launch? The presets idea I had was more along the lines of moving a group of user specified apps on demand with preset buttons. The use case for this would be if a user had headphones connected but needed to show someone a video or a song. They could make a preset button to move only their browser and spotify to the device speakers and then toggle it back.
4. For the Windows 11 style changes, would you prefer if they could be toggled for windows 10 users? Because right now my implementation just has them baked in.

**Reply by D-Arora** [reply]
Hi @riverar or other devs, any feedback on these questions? Thanks!

**Reply by riverar** [reply]
@D-Arora Haven't forgotten about you. Let's see.

1. We considered drag/drop but at the time it was thought to be very complex to support. There's a lot of design work that has to go into it -- Should this support touch? Is it accessible? Does RTL or high DPI have any impact to this feature? Do the items slide around as you drag or is there a divider line that appears? What if the flyout is collapsed? What if it's too tall? Where do folks grab the row to drag? Nearly all the surfaces are either buttons or click/touch targets...

   It's not off the table though so if you'd like to think through all that and propose something, that'd be great.

2. You can rename renderers in Windows today (via the legacy Sounds dialog). But yeah, agree, would be nice to support full rename. Both the display name and device name are properties that should be settable, so the backend is easy. The harder part is frontend + localization.

3. Install a dev build and it'll appear in EarTrumpet Settings. You then have an IFTTT-like grammar to work with. It's easier to not try to explain and just encourage you to try it out. 🙃 

4. Windows 11 style changes should only be visible for Windows 11 users. Windows 10 users should not be able to enable those styles. We have a pretty robust theming component (see dev docs) that should make this easy. I recently added Windows 11 OS detection to it as well.

**Comment by Baxmann** [comment]
Since I find it semi-related, I would like to bring up the suggestion to [change right click to match Windows 11](https://github.com/File-New-Project/EarTrumpet/discussions/1039).

**Comment by diego-gz** [comment]
I also have the similiar idea, the icon on the taskbar is also kinda odd if put beside other windows 11 icons & dropdown
![image](https://user-images.githubusercontent.com/83010300/170807349-9c90e43c-5e1e-4ca6-9679-024f732fbe11.png)
